packages/0xcert-ethereum-gateway-contracts/src/contracts/token-deploy-gateway.sol

```diff
@@ -0,0 +1,411 @@
+   * or listen to Cancel event if you want to check if an deploy is already canceled.
+   * @param _data Data of deploy to cancel.
+   */
+  function cancel(
+    DeployData memory _data
+  )
+    public
+  {
+    require(_data.maker == msg.sender, MAKER_NOT_EQUAL_TO_SENDER);
+
+    bytes32 claim = getDeployDataClaim(_data);
+    require(!deployPerformed[claim], DEPLOY_ALREADY_PERFORMED);
+
+    deployCancelled[claim] = true;
+    emit Cancel(
+      _data.maker,
+      _data.taker,
+      claim
+    );
+  }
+
+  /**
+   * @dev Calculates keccak-256 hash of DeployData from parameters.
+   * @param _deployData Data needed for atomic swap.
+   * @return keccak-hash of deploy data.
+   */
+  function getDeployDataClaim(
+    DeployData memory _deployData
+  )
+    public
+    view
+    returns (bytes32)
+  {
+    bytes32 tokendata = keccak256(
+      abi.encodePacked(
+        _deployData.tokenData.name,
+        _deployData.tokenData.symbol,
+        _deployData.tokenData.supply,
+        _deployData.tokenData.decimals,
+        _deployData.tokenData.owner
+      )
+    );
+
+    bytes32 transferData = keccak256(
+      abi.encodePacked(
+        _deployData.transferData.token,
+        _deployData.transferData.to,
+        _deployData.transferData.value
+      )
+    );
+
+    return keccak256(
+      abi.encodePacked(
+        address(this),
+        _deployData.maker,
+        _deployData.taker,
+        tokendata,
+        transferData,
+        _deployData.seed,
+        _deployData.expiration
+      )
+    );
+  }
+
+  /**
+   * @dev Verifies if claim signature is valid.
+   * @param _signer address of signer.
+   * @param _claim Signed Keccak-256 hash.
+   * @param _signature Signature data.
+   */
+  function isValidSignature(
+    address _signer,
+    bytes32 _claim,
+    SignatureData memory _signature
+  )
+    public
+    pure
+    returns (bool)
+  {
+    if (_signature.kind == SignatureKind.eth_sign)
+    {
+      return _signer == ecrecover(
+        keccak256(
+          abi.encodePacked(
+            "\x19Ethereum Signed Message:\n32",
+            _claim
+          )
+        ),
+        _signature.v,
+        _signature.r,
+        _signature.s
+      );
+    } else if (_signature.kind == SignatureKind.trezor)
+    {
+      return _signer == ecrecover(
+        keccak256(
+          abi.encodePacked(
+            "\x19Ethereum Signed Message:\n\x20",
+            _claim
+          )
+        ),
+        _signature.v,
+        _signature.r,
+        _signature.s
+      );
+    } else if (_signature.kind == SignatureKind.eip712)
+    {
+      return _signer == ecrecover(
+        _claim,
+        _signature.v,
+        _signature.r,
+        _signature.s
+      );
+    }
+
+    revert(INVALID_SIGNATURE_KIND);
+  }
+
+  /**
+   * @dev Helper function that makes deploy actions.
+   * @param _deploy Data needed for deploy.
+   */
+  function _doActions(
+    DeployData memory _deploy
+  )
+    private
+    returns (address _token)
```

Comment:
I see no benefit of specifying this as an ERC20 address. It is only used to emit in event where it is specified as address so either way it will go back to plain address.